src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapterProvider.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using NuGet.ProjectManagement;
+using NuGet.VisualStudio;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    [Export(typeof(IVsProjectAdapterProvider))]
+    internal class VsProjectAdapterProvider : IVsProjectAdapterProvider
+    {
+        private readonly IDeferredProjectWorkspaceService _workspaceService;
+        private readonly IVsProjectThreadingService _threadingService;
```

Comment:
lazy import? since DeferredProjectWorkspaceService should only be initialized when solution is deferred load.